src/util/Signing.ts

```diff
@@ -0,0 +1,136 @@
+/*
+	Spacebar: A FOSS re-implementation and extension of the Discord.com backend.
+	Copyright (C) 2023 Spacebar and Spacebar Contributors
+	
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published
+	by the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+	
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+	
+	You should have received a copy of the GNU Affero General Public License
+	along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { Config } from "@spacebar/util";
+import { createHmac, timingSafeEqual } from "crypto";
+import ms, { StringValue } from "ms";
+import { ParsedQs } from "qs";
+
+export const getUrlSignature = (path: string) => {
+	const { cdnSignatureKey, cdnSignatureDuration } = Config.get().security;
+
+	// calculate the expiration time
+	const now = Date.now();
+	const issuedAt = now.toString(16);
+	const expiresAt = (now + ms(cdnSignatureDuration as StringValue)).toString(
+		16,
+	);
+
+	// hash the url with the cdnSignatureKey
+	const hash = createHmac("sha256", cdnSignatureKey as string)
+		.update(path)
+		.update(issuedAt)
+		.update(expiresAt)
+		.digest("hex");
```

Comment:
No origin IP address/other requester info? Media shouldn't be available to sharing targets.

Reply:
Why not? This is the same behaviour as Discord, afaik